server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1218,6 +1306,29 @@ private void finalizeSnapshotEntry(SnapshotsInProgress.Entry entry, Metadata met
         }
     }
 
+    /**
+     * Removes all feature states which have missing or failed shards, as they are no longer safely restorable.
+     * @param entry The "in progress" entry with a list of feature states and one or more failed shards.
+     * @return The list of feature states which were completed successfully in the given entry.
+     */
+    private List<SnapshotFeatureInfo> onlySuccessfulFeatureStates(SnapshotsInProgress.Entry entry) {
+        assert entry.partial() : "should not try to filter feature states from a non-partial entry";
+
+        // Figure out which indices have unsuccessful shards
+        Set<String> indicesWithUnsuccessfulShards = new HashSet<>();
+        entry.shards().keysIt().forEachRemaining(shardId -> {
+            final ShardState shardState = entry.shards().get(shardId).state();
+            if (shardState.failed() || shardState.completed() == false) {
+                indicesWithUnsuccessfulShards.add(shardId.getIndexName());
+            }
+        });
+
+        // Now remove any feature states which contain any of those indices, as the feature state is not intact and not safely restorable
+        return entry.featureStates().stream()
```

Comment:
I think you need to do this the other way around. If you have a partial snapshot, the indices just disappear from the `Entry` instead of showing up as failure. So if indices are missing, this will just quietly ignore those. You can just run the logic the other way around and check that all the indices you need for a feature are in the indices list.

Reply:
I'm not entirely sure what you mean here - I tried replacing this method with:
```java
final Set<String> indicesInSnapshot = entry.indices().stream()
    .map(IndexId::getName)
    .collect(Collectors.toUnmodifiableSet());

return entry.featureStates().stream()
    .filter(stateInfo -> indicesInSnapshot.containsAll(stateInfo.getIndices()))
    .collect(Collectors.toList());
```

And the test I added to check partial snapshots failed - the index which was only partially captured was still in `entry.indices()`. What I did was apply both approaches (see [this commit](https://github.com/elastic/elasticsearch/pull/63513/commits/38022ddf0fb242b40f239c7398e887f2ee05edf0)) - check for any indices which had failed shards and remove any feature states which include those indices AND check for any indices in feature states which aren't in `entry.indices()`, but I couldn't figure out how to force that case in a test.

Is that what you mean? Or did I misunderstand you?

Reply:
@gwbrown 

> Is that what you mean?

Yea exactly we gotta check both. Sorry I missed that you did that correctly in that commit :) Looks good.
Can we still add a test for this please?

> check for any indices in feature states which aren't in entry.indices(), but I couldn't figure out how to force that case in a test.

This should be easy:

1. Start `partial=true` snapshot and block it somewhere
2. Wait for block
3. Delete index
4. Let snapshot continue
5. The finalisation should go through fine and and the feature should not become part of the snapshot.

That should work?

Reply:
@original-brownbear 

I've written a test that follows that pattern, and it passes, but I'm concerned because if I comment out this line, which applies the logic you recommended of checking `entry.indices()`:
https://github.com/elastic/elasticsearch/blob/90bf18dfa85e4c2a8c896e914d56062ee42ea56c/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java#L1354

...then the test *still passes*, which it sounds like it shouldn't. As far as I can tell, when this test runs, the `featureStates` field in the initial `Entry` doesn't get populated until after the index is deleted, so that feature state is never included in entry to begin with.

I've tried blocking on data nodes instead of blocking the master (see the commented-out lines [removed in this commit](https://github.com/elastic/elasticsearch/pull/63513/commits/d3005866c05e13a7ec7df20824e1680b39a737b9)), and that version also passes, and again still passes if I comment out the same line - in that case, at least one shard fails and the feature state is removed via the logic I originally had in place.

I also tried using, instead of `blockNodeOnAnyFiles`:
- `blockMasterOnWriteIndexFile`, which results in the snapshot completing with all shards successful
- `blockMasterFromFinalizingSnapshotOnSnapFile`, which results in the snapshot entirely failing (`SnapshotMissingException` when attempting to get the `SnapshotInfo`)
- `blockMasterFromFinalizingSnapshotOnIndexFile`, which has the same results as `blockMasterFromFinalizingSnapshotOnSnapFile`

So I'm still confused about how to actually get the original situation you describe, where we can get a feature state in the `Entry` that references an index not in `entry.indices()`.

If you have bandwidth, I'd be happy to talk synchronously on this.

Reply:
I talked with @original-brownbear synchronously and figured out what was going on. It turned out that with some recent snapshot optimizations, it's more difficult to trigger the case he originally described.

In order to trigger the case reliably, the test now:
- After creating the system indices, starts a new data node
- Creates an index allocated only on the new data node
- Blocks the new data node and starts a snapshot
- Waits for the block to be hit
- Deletes the index
- Unblocks the data node

This allows all shards of the system index to be snapshotted, so that there are no failed shards, but still have that index removed from the snapshot.

There was also something else I misunderstood - when the index is deleted, it's removed from the final `SnapshotInfo`, but not the `Entry` - the final list of indices is generated via `ShardGenerations`. So the method to filter feature states now takes the final list of indices as well as the `Entry` and checks against that, rather than against `Entry#indices()` as the latter does not change.